oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+                        viewData.addMessage(Message(LocalDateTime.now(), result.name, "Evaluation try to cancel Timed out after ${sim.timeOut!!.toMillis()} millisecond but Failed: Due to\n${result.exception}"))
+                        viewData.addResult(Result(result.name, "Cancel Failed", inputVector.toString(), "Cancellation failed:\n${result.exception}"))
+                    }
+                }
+                i++
+            }
+        }
+        viewData.updateStatusMessage("Idle")
+    }
+
+
+    fun stopOptimization() {
+        state = State.Idle
+    }
+
+    fun syncAll() = GlobalScope.launch {
+        for (simName in simulationsByName.keys) {
+            syncConfigFor(simName)
+        }
+    }
+
+
+    fun cancelAll() = GlobalScope.launch {
+        for ((name, sim) in simulationsByName) {
+            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+            ).build()
+
+            sim.input.onNext(message)
+        }
+    }
+
+    fun cancel(name: String) = GlobalScope.launch {
+        val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+        ).build()
+
+        simulationsByName.getValue(name).input.onNext(message)
+    }
+
+    fun setDuration(nodeName: String?, timeOut: Duration?) {
+        if (nodeName == null || !simulationsByName.containsKey(nodeName)) {
+            return
+        }
+        val node: Simulation = simulationsByName.getValue(nodeName)
+        val newNode = node.copy(
+                timeOut = timeOut
+        )
+        simulationsByName += nodeName to newNode
+    }
+
+    fun disconnectAll() = GlobalScope.launch {
+        for ((name, sim) in simulationsByName) {
+            try {
+                sim.input.onCompleted()
+            } catch (e: StatusRuntimeException) {
+                println("Error when close input for:\n$e")
+            }
+        }
+    }
+
+    fun cancelAndStop() {
+        GlobalScope.launch {
+            stopOptimization()
+            for ((name, sim) in simulationsByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    private suspend fun syncConfigFor(simName: String) {
+        val sim = simulationsByName.getValue(simName)
+        val message = OASISQueryDTO.newBuilder().setNodeStatusRequest(
+                OASISQueryDTO.NodeStatusUpdateRequest.newBuilder().setName(simName)
+        ).build()
+
+        sim.input.onNext(message)
+
+        val result = select<Either<Simulation, Message>> {
+            sim.update.onReceive { it -> Either.Left(updateFromResponse(it)) }
+            sim.error.onReceive { it ->
```

Comment:
Hmm, so I'm a little worried about the multiplexing happening here. 

What if two functions both call `sim.error.onRecieve` at the same time? Any idea who gets the error?

Reply:
Do you mena if we have two selects on the same channel?

Reply:
this is talked, we will have new issue to address this